Headless backend

emersion opened the pull request:
Test plan: launch rootston, open `gnome-calculator`, take a screenshot.

Fixes #495

SirCmpwn commented on backend/headless/backend.c:
```diff
@@ -0,0 +1,166 @@
+		(struct wlr_headless_backend *)wlr_backend;
+	wlr_log(L_INFO, "Initializating headless backend");
+
+	// TODO
```
?

emersion replied:
This `i < 1` and `1280, 720` need to be replaced. Maybe I'll just refuse to create outputs before backend startup, and do not create any output by default.

SirCmpwn replied:
Should probably cache output create requests like Wayland does I suppose.

emersion replied:
An issue with this is that NULL will be returned by `wlr_headless_add_output` instead of a `wlr_output`.

SirCmpwn replied:
Nah. It's not like backend_start is really doing anything here. You already have an EGL context ready once init is called.

emersion replied:
Okay, so let's make `wlr_headless_add_output` create outputs regardless of whether or not the backend is started, and start output timers on backend startup.

Should I use one timer for all outputs, or one timer per output like it's done currently?

SirCmpwn replied:
One timer per output and make the framerate configurable. Let's also make the resolution reconfigurable at runtime.

emersion commented:
This PR is ready for review. Since backends are not configurable in rootston, there are some changes in `main.c` that'll need to be reverted before merge.

SirCmpwn commented:
Tested, works great. Great work!

SirCmpwn commented:
Code looks good too. :+1:

ascent12 commented on backend/headless/output.c:
```diff
@@ -0,0 +1,146 @@
+}
+
+static void output_swap_buffers(struct wlr_output *wlr_output) {
+	// No-op
```
Are you sure you don't need to call eglSwapBuffers here?

emersion replied:
From https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml

>If surface is a pixel buffer or a pixmap, eglSwapBuffers has no effect, and no error is generated.

ascent12 commented:
There is an awful lot of duplication of the EGL code.
I feel the code in render/egl.c could be made more generally useful instead of just copying it.

SirCmpwn commented:
I thought the same thing, but the changes touch a few places and the code isn't that long so I didn't object. I wouldn't mind seeing it refactored, though.

ascent12 commented:
Would calling `wlr_egl_init(&backend->egl, EGL_PLATFORM_SURFACELESS_MESA)` not work for this?
It would require some changes to how EGLConfigs are handled, though. We'd need some way to get custom attributes to it, and possibly separate it out from `wlr_egl_init` entirely, and make it its own API function.

SirCmpwn commented:
That should basically be it, yes.

emersion commented:
I didn't want to split `wlr_egl_init` into two functions and allow users to half-initialize an `wlr_egl`, so I added a `config_attribs` parameter.

`eglGetPlatformDisplayEXT` isn't working with `EGL_PLATFORM_SURFACELESS_MESA` (even if the docs say it should, tried `eglGetPlatformDisplayEXT(EGL_PLATFORM_SURFACELESS_MESA, EGL_DEFAULT_DISPLAY, NULL)`), so I added a special case for this one.

SirCmpwn commented:
Thanks!

martinetd commented on rootston/main.c:
```diff
@@ -29,7 +30,11 @@ int main(int argc, char **argv) {
+	assert(server.backend = wlr_headless_backend_create(server.wl_display));
+	wlr_headless_add_output(server.backend, 1280, 720);
+	wlr_headless_add_input_device(server.backend, WLR_INPUT_DEVICE_KEYBOARD);
+	wlr_headless_add_input_device(server.backend, WLR_INPUT_DEVICE_POINTER);
```
Shouldn't this part be commented out/reverted for now?
Not sure if it does any harm but it might confuse output lists/things like that.
